### PR TITLE
WIP/CI/DOC: Build docs in CircleCI and create preview link

### PR DESCRIPTION
**Pull request by datapythonista**
Trying the same approach to generating the preview of the docs as numpy/scipy/matplotlib, using CircleCI and an action to find the link as a CI job status.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1,6 +1,20 @@
 version: 2.1
 
 jobs:
+  build-docs:
```
Comment (mroeschke) [inline .circleci/config.yml:4]: I think you may need to add this to `workflows` below

File: .github/workflows/doc-previewer.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    permissions:
+      statuses: write
+    if: "${{ github.event.context == 'ci/circleci: build-docs' }}"
+    steps:
+      - name: CircleCI artifact redirector
+        uses: larsoner/circleci-artifacts-redirector-action@master
+        with:
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          artifact-path: 0/web/build/index.html
+          circleci-jobs: build-docs
```
Comment (melissawm) [inline .github/workflows/doc-previewer.yml:17]: You can also use `job-title` to choose the wording for the preview status, e.g.

```
job-title: Check the rendered docs here!
```

**Comment by pmeier** [comment]
I've implemented a PoC for 1./2. in #52139.

**Comment by datapythonista** [comment]
Not sure what's going on here, the error is not very descriptive. Feels like maybe the CircleCI worker may not have enough resources to build our docs?

@lithomas1 any idea?

**Comment by lithomas1** [comment]
> Not sure what's going on here, the error is not very descriptive. Feels like maybe the CircleCI worker may not have enough resources to build our docs?
> 
> @lithomas1 any idea?

We have a 1hr timeout, can you try to build the docs in parallel?

**Comment by datapythonista** [comment]
> We have a 1hr timeout, can you try to build the docs in parallel?

Our docs should be with as many processes as cores by default.

From the logs, I don't know if the problem is the timeout, or if something crashes and the process never finishes...

**Comment by lithomas1** [comment]
The duration is 1h 0m 1s, it must be the timeout I think.

I don't think there is a way around it (other than emailing CircleCI and asking them to raise our timeout).

**Comment by lithomas1** [comment]
You could also try specifying the large resource class, which should give 4 cpu. Right now, I think it only uses the medium class, which has 2.

**Comment by datapythonista** [comment]
> You could also try specifying the large resource class, which should give 4 cpu. Right now, I think it only uses the medium class, which has 2.

I still don't think the problem is the timeout, I think the docs process is crashing, but good idea, as maybe the problem is memory. Let's see how it goes.

**Comment by datapythonista** [comment]
Seems like we continue to get this kind of traceback / profiling info that we shouldn't get after the workers start, something seems to be broken, but impossible to tell what based on this:
```
         618142 function calls (618124 primitive calls) in 0.137 seconds

   Ordered by: internal time
   List reduced from 182 to 4 due to restriction <4>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1000    0.079    0.000    0.118    0.000 <ipython-input-4-c2a74e076cf0>:1(i ntegrate_f)
   552423    0.038    0.000    0.038    0.000 <ipython-input-3-c138bdd570e3>:1(f )
     3000    0.004    0.000    0.014    0.000 series.py:970(__getitem__)
     3000    0.002    0.000    0.006    0.000 series.py:1076(_get_value)
```

I'll wait to see if the job fails as I expect, and will close this when it does. I think the current system based on our own server is working fine, and no need to move it to CircleCI if it's not easy.

**Comment by datapythonista** [comment]
No idea what's going on and why CircleCI can't build our docs. I don't think it's worth spending more time on this, and in particular if it doesn't seem reliable and able to provide useful errors on failures. Closing.

**Pull request closed without merging**